buildinfo: named input contexts support

crazy-max opened the pull request:
follow-up https://github.com/moby/buildkit/pull/2476

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
Second example https://github.com/docker/buildx/blob/master/docs/reference/buildx_bake.md#using-a-secondary-source-directory

```Dockerfile
# Dockerfile

FROM scratch AS src

FROM busybox
COPY --from=src . /src
RUN ls -al /src
```

```hcl
# docker-bake.hcl
target "app" {
  contexts = {
    src = "../ctx-src"
  }
}
```

```console
$ docker buildx bake --metadata-file metadata.json --set *.args.foo=bar --set *.args.bar=foo app
$ cat metadata.json
```
```json
{
  "containerimage.buildinfo": {
    "frontend": "dockerfile.v0",
    "attrs": {
      "add-hosts": "",
      "build-arg:bar": "foo",
      "build-arg:foo": "bar",
      "context:src": "local:src",
      "filename": "Dockerfile",
      "frontend.caps": "moby.buildkit.frontend.contexts+forward"
    },
    "sources": [
      {
        "type": "docker-image",
        "ref": "busybox",
        "alias": "docker.io/library/busybox@sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb",
        "pin": "sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb"
      }
    ]
  }
}
```

`add-hosts` should not be displayed if empty and `frontend.caps` should be filtered out I think.

crazy-max commented:
Hum actually looking at all the examples result the filter logic looks wrong. Will take a look.

tonistiigi commented:
Deps map should not contain buildinfo for the sources, read from the config. The goal is not to gather all build sources recursively(and atm it is just one level of recursiveness anyway). Only for the buildinfo sent with inputs in https://github.com/docker/buildx/pull/958 . The difference is that while for image sources we can just as buildinfo from the image itself, for the inputs they don't exist anywhere. So if you rebuild from buildinfo you need to rebuild the deps as well and therefore need buildinfo for the deps.

I'm not sure how this looks like for multi-arch builds as well. I think atm the platform would appear in these keys and I'm not sure if that is correct or not. Feels wrong if you get different buildinfo depending on if you built for a single arch or for the one arch inside of the multi-arch image.

crazy-max commented:
@tonistiigi 

> Deps map should not contain buildinfo for the sources, read from the config. The goal is not to gather all build sources recursively(and atm it is just one level of recursiveness anyway). Only for the buildinfo sent with inputs in https://github.com/docker/buildx/pull/958 . The difference is that while for image sources we can get buildinfo from the image itself, for the inputs they don't exist anywhere. So if you rebuild from buildinfo you need to rebuild the deps as well and therefore need buildinfo for the deps.

Yes was thinking it could be useful but that's out of scope agree and should be handled by others tooling. I removed that behavior.

> I'm not sure how this looks like for multi-arch builds as well. I think atm the platform would appear in these keys and I'm not sure if that is correct or not. Feels wrong if you get different buildinfo depending on if you built for a single arch or for the one arch inside of the multi-arch image.

I've added a test for multiplatform and deps https://github.com/moby/buildkit/pull/2654/files#diff-7df700366904727b0358fe2ed234172b34a8d68782c1ef1c255fb8379db79c9bR385. Excpected output for buildinfo:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base::linux/amd64": "input:base::linux/amd64",
    "context:base::linux/arm64": "input:base::linux/arm64"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
      "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    }
  ],
  "deps": {
    "base:linux/amd64": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "alpine",
          "alias": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    },
    "base:linux/arm64": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "alpine",
          "alias": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

tonistiigi commented:
```
{
  "app": {
    "containerimage.buildinfo/linux/amd64": {
      "frontend": "dockerfile.v0",
      "attrs": {
        "build-arg:foo": "bar",
        "context:baseapp": "target:base",
        "context:baseapp::linux/amd64": "input:0-base::linux/amd64",
        "context:baseapp::linux/arm64": "input:0-base::linux/arm64",
        "filename": "Dockerfile"
      },
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/busybox@sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb",
          "pin": "sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb"
        }
      ],
      "deps": {
        "0-base::linux/amd64": {
          "frontend": "dockerfile.v0",
          "attrs": {
            "filename": "baseapp.Dockerfile"
          },
          "sources": [
            {
              "type": "docker-image",
              "ref": "busybox",
              "alias": "docker.io/library/busybox@sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb",
              "pin": "sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb"
            }
          ]
        },
        "0-base::linux/arm64": {
          "frontend": "dockerfile.v0",
          "attrs": {
            "filename": "baseapp.Dockerfile"
          },
          "sources": [
            {
              "type": "docker-image",
              "ref": "busybox",
              "alias": "docker.io/library/busybox@sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb",
              "pin": "sha256:afcc7f1ac1b49db317a7196c902e61c6c3c4607d63599ee1a82d702d249a0ccb"
            }
          ]
        }
      }
    },
```

Is what I see from bake atm. Under `containerimage.buildinfo/linux/amd64` there should only be amd64 contexts/deps. Also `"context:baseapp": "target:base",` should either not be there or have the platform-specific value.

Ideally, we can clean up the names like `input:0-base` as well, but I think that is buildx side change.

crazy-max commented:
> Is what I see from bake atm. Under `containerimage.buildinfo/linux/amd64` there should only be amd64 contexts/deps. Also `"context:baseapp": "target:base",` should either not be there or have the platform-specific value.

Now platforms not tied to the metadata key (eg `containerimage.buildinfo/linux/amd64`) are filtered out for `attrs` and `deps`:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
      "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "alpine",
          "alias": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

> Ideally, we can clean up the names like `input:0-base` as well, but I think that is buildx side change.

Yes it should be done on buildx.

tonistiigi commented on util/buildinfo/buildinfo.go:
```diff
@@ -136,6 +143,60 @@ func mergeSources(ctx context.Context, buildSources map[string]string, frontendS
+
+		// if platform is defined, only decode dependencies for that platform
+		if platform != "" && !strings.HasSuffix(k, "::"+platform) {
+			continue
```
As discussed offline, we should do a follow-up that handles the contexts without suffixes as well for multi-platform cases.
